lib/spack/spack/stage.py

```
@@ -725,6 +728,191 @@ def _add_to_root_stage(self):
                     install(src, destination_path)
 
 
+class CMakeBuildStage:
+    """Interface abstracting a CMake build tree at a location outside
+    of a Spack stage directory but still managed by Spack.
+
+    CMake's build tree can be located arbitrarily on a filesystem independent
+    from a source directory. This class relocates that build directory out of the
+    stage and to a path under a Users home directory or to a location of a users chosing set
+    either by config or the command line via the --cmake-build-dir cl argument
+    to the install command.
+    Interactions with the stage expecting an in stage build tree will work as normal
+    as this class serves to obfucscate the external stage and allow for all stage behavior
+    to perform as normal.
+    The external build tree is given the same lifespan as its stage dir, is spun up when the stage
+    spins up, and destroyed when the stage is destroyed. After the build and installation are done,
+    the build tree is first relocated to what would be its proper place in the stage.
+    and a symlink is placed in place of the build tree pointing at the external directory
+
+    Note: This is not, nor should it be, used on *nix platforms and is intended as a solution to
+    reduce file path lengths on Windows during compilation/linking. This class should be removed
+    when MSVC fully supports the LongPath feature on Windows.
+    """
+
+    dispatch: Dict[str, str] = {}
+
+    def __init__(self, hash, name, root=None, keep=False):
+        # Users can override external cmake build dir, default is %USERPROFILE%
+        # overrides can come from command line or config, command line will override all
+        self._hash = hash
+        self._path = Path(get_stage_root(), name)
+        self._remote_stage = None
+        self.keep = keep
+        if not root:
+            fallback_path = Path(os.environ["USERPROFILE"], ".sp-stage")
```

I'd prefer not having a fallback: raise an error if the alternative root is not configured. I'm trying to limit automatic usage of `~` (or an analog of it on Windows).

---

Sounds good. In that case I'll remove the other config option and just have the behavior enabled by whether or not a user sets the external path in the config.